### PR TITLE
Use weakref to avoid cache leaks in fori_loop.

### PR DESCRIPTION
**Pull request by pschuh**
This is in response to: https://github.com/google/jax/issues/11448.

### CODE REVIEW

**Comment by YouJiacheng** [comment]
FWIW, similar problems:
https://github.com/google/jax/issues/9714
https://github.com/google/jax/issues/10023#issuecomment-1077812932
Could this PR also close #9714 ?

**Review by jakevdp — APPROVED** [review]
Looks fine - is it fair to say that we should use `weakref_lru_cache` in place of `cache()` for anything that takes a function as its primary argument?

**Comment by pschuh** [comment]
@jakevdp yea, I think that is reasonable. It is also good for caches that take jaxprs as the first argument.

File: jax/_src/lax/control_flow/loops.py [hunk 1]
```diff
@@ -1497,18 +1499,20 @@ def _fori_cond_fun(loop_carry):
...
 
-@cache()
+@weakref_lru_cache
 def _fori_scan_body_fun(body_fun):
+  body_fun = weakref.ref(body_fun)
   def scanned_fun(loop_carry, _):
     i, x = loop_carry
-    return (i + 1, body_fun(i, x)), None
+    return (i + 1, body_fun()(i, x)), None
   return scanned_fun
```
Comment (YouJiacheng) [inline jax/_src/lax/control_flow/loops.py:1516]: IIUC, these two functions are not expensive at all, why we need a cache here?

**Comment by YouJiacheng** [comment]
https://github.com/google/jax/blob/86ab8a84ee9c0c1fc718f3d322f7f597bebb812b/jax/_src/lax/control_flow/loops.py#L1502-L1516
IIUC, these two functions are not expensive at all, why we need cache here?

**Pull request merged**